### DISCUSSION TITLE
net-net example: unknown interface version `'

### DISCUSSION THREAD

**Discussion by angelofloridia** [strongswan/strongswan#1612] [Q&A] [answered]
Hello everyone,
I'm trying to replicate the net-net with auto start example ([link](https://www.strongswan.org/testing/testresults/ikev2/net2net-start/)).
The network address are a little bit different, but should change nothing.

```
10.0.0.0/24 -- | 192.168.57.3 | === | 192.168.57.7 | -- 10.0.1.0/24
  moon-net          moon                 sun           sun-net
```

I'm using virtualbox, Moon and Sun has two network interfaces, one is host-only and one is an internal network to connect with Alice and Bob respectively.

I added a default route to Alice and Bob, so that they will use Moon and Sun to ping each other.

This is the swanctl.conf file in Moon, similar is in Sun of course.
```
connections {
   gw-gw {
      local_addrs  = 192.168.57.3
      remote_addrs = 192.168.57.7
      local {
         auth = pubkey
         certs = moonCert.pem
         id = moon.strongswan.org
      }
      remote {
         auth = pubkey
         id = sun.strongswan.org
      }
      children {
         net-net {
            local_ts = 10.0.0.0/24
            remote_ts = 10.0.1.0/24
            start_action = start 
            updown = /usr/libexec/ipsec/_updown iptables
            rekey_time = 5400
            rekey_bytes = 500000000
            rekey_packets = 1000000
            esp_proposals = aes128-sha2_256-modp2048
         }
      }
      version = 2
      mobike = no
      reauth_time = 10800
   }
}
```
Now, from what I get from the documentation, the updown script should set all the iptables configuration. I suspect it doesn´t but don't get why.

my `iptables -L` ouput
```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         
ACCEPT     all  --  10.0.1.0/24          10.0.0.0/24          policy match dir in pol ipsec reqid 1 proto esp
ACCEPT     all  --  10.0.0.0/24          10.0.1.0/24          policy match dir out pol ipsec reqid 1 proto esp

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
```

example `iptables -L` ouput
```
=== filter table ===
Chain INPUT (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    1   140 ACCEPT     esp  --  eth0   *       0.0.0.0/0            0.0.0.0/0           
    0     0 ACCEPT     ah   --  eth0   *       0.0.0.0/0            0.0.0.0/0           
    3  2189 ACCEPT     udp  --  eth0   *       0.0.0.0/0            0.0.0.0/0            udp spt:500 dpt:500
    0     0 ACCEPT     udp  --  eth0   *       0.0.0.0/0            0.0.0.0/0            udp spt:4500 dpt:4500
  101 23496 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:22
    4  1108 ACCEPT     tcp  --  eth0   *       192.168.0.150        0.0.0.0/0            tcp spt:80

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    1    84 ACCEPT     all  --  eth0   *       10.2.0.0/16          10.1.0.0/16          policy match dir in pol ipsec reqid 1 proto 50
    1    84 ACCEPT     all  --  *      eth0    10.1.0.0/16          10.2.0.0/16          policy match dir out pol ipsec reqid 1 proto 50

Chain OUTPUT (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    1   140 ACCEPT     esp  --  *      eth0    0.0.0.0/0            0.0.0.0/0           
    0     0 ACCEPT     ah   --  *      eth0    0.0.0.0/0            0.0.0.0/0           
    3  2236 ACCEPT     udp  --  *      eth0    0.0.0.0/0            0.0.0.0/0            udp spt:500 dpt:500
    0     0 ACCEPT     udp  --  *      eth0    0.0.0.0/0            0.0.0.0/0            udp spt:4500 dpt:4500
  116 25488 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp spt:22
    6   391 ACCEPT     tcp  --  *      eth0    0.0.0.0/0            192.168.0.150        tcp dpt:80

=== nat table ===
Chain PREROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

=== mangle table ===
Chain PREROUTING (policy ACCEPT 279 packets, 90769 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain INPUT (policy ACCEPT 277 packets, 90601 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain FORWARD (policy ACCEPT 2 packets, 168 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 323 packets, 57315 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain POSTROUTING (policy ACCEPT 325 packets, 57483 bytes)
 pkts bytes target     prot opt in     out     source               destination         
```

What am I missing?

Also, the output of `/usr/libexe/ipsec/_updown iptables` is:
```bash
moon@moon:/etc/strongswan.d$ sudo /usr/libexec/ipsec/_updown iptables
/usr/libexec/ipsec/_updown: unknown interface version `'
```

Of course the ping between Alice and Bob does not work, because Moon and Sun does not have the right configuration on where to forward packets.

Can anyone help? 

Thanks in advance!

EDIT: As written in documentation 

> If the local traffic selector is not a single host then the option
> 
>` connections.<conn>.children.<child>.hostaccess = yes `
> 
> in [swanctl.conf](https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html) inserts rules in the INPUT and OUTPUT chains, besides the rules in the FORWARD chain, that allow accessing the VPN server itself. Please refer to the _updown script for details.

So I added the row in my swanctl.conf file and now the configuration of iptables seems pretty much ok, but still it does not work! 
When I ping Bob from Alice, Moon still does not know how to reach the destination network.

This is the new iptables -L output:
```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         
ACCEPT     all  --  10.0.1.0/24          10.0.0.0/24          policy match dir in pol ipsec reqid 1 proto esp

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         
ACCEPT     all  --  10.0.1.0/24          10.0.0.0/24          policy match dir in pol ipsec reqid 1 proto esp
ACCEPT     all  --  10.0.0.0/24          10.0.1.0/24          policy match dir out pol ipsec reqid 1 proto esp

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
ACCEPT     all  --  10.0.0.0/24          10.0.1.0/24          policy match dir out pol ipsec reqid 1 proto esp

```

I know at this point I may be missing something very little but can´t figure out yet

**Comment by tobiasbrunner** [comment, marked as answer]
> ow, from what I get from the documentation, the updown script should set all the iptables configuration.

Note that these rules are only necessary if the firewall is configured to block traffic. With your `ACCEPT` default policies for all chains they are not necessary (as opposed to the example, where you can see the `DROP` default policies).







> Also, the output of `/usr/libexe/ipsec/_updown iptables` is:
> 
> ```shell
> moon@moon:/etc/strongswan.d$ sudo /usr/libexec/ipsec/_updown iptables
> /usr/libexec/ipsec/_updown: unknown interface version `'
> ```

Yeah, you can't call it just like that as it depends on a lot of environment variables that are set by the [_updown_ plugin](https://docs.strongswan.org/docs/5.9/plugins/updown.html) when it calls the script.

> Of course the ping between Alice and Bob does not work, because Moon and Sun does not have the right configuration on where to forward packets.

What configuration are you referring to?  Anyway, did you check the [docs regarding forwarding](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html)?

**Comment by angelofloridia** [comment]
Thanks for answering! And also thanks for the explanation about iptables, that's new for me.

I had checked the docs about forwarding but now I read it very carefully.
In my case, which is a site-to-site case, I should configure on the hosts in the subnets a route to reach the other subnet via the VPN gateway.
Now Alice (10.0.0.4 in moon subnet) is configured as follow (Bob is configured respectively): 
```
alice@alice: ~$ sudo ip route
default via 10.0.0.1 #the default gateway and the vpn gateway
10.0.0.0/24 dev enp0s3 via proto kernel scope link src 10.0.0.4
10.0.1.0/24 dev via 10.0.0.1 dev enp0s3
```

What I see from tcpdump on moon is that when alice ping bob the packet goes to Moon gateway but the latter answer back `Destination Net unreachable`.
This is what makes me think the gateways are not correctly configured.

I edited the `/etc/sysctl.conf` file to enable forwarding and I still have the issue.

Any other recommendation? Thanks again for answering in the first place, this is very helpful!

EDIT:
I managed to solve the problem!
As suggested I don't need the updown plugin so I removed from the sawnctl.conf file.
Also I configured static routes on the gateways which, together with the forwarding rules active, made it works! 

Thanks a bunch, Strongswan is an amazing tool!

**Comment by angelofloridia** [comment]
Hi all, I don't know if it wise to continue this thread, in case not I'll be happy to open another one.

The configuration is the same as the first message, everything works like a charm and I did a lot of tests now.

I'm currently focusing on executing the daemon with reduced privileges.

What I did:
1. I build strongswan using `--with-capabilities=native ` (I'm on Linux ofc)
2. Edited the `/etc/strongswan.conf` file and added the row `user=myuser ` to be able to execute the `swanctl` tool as non-root
3. `swanctl --load-all`
4. 'swanctl --initiate --child net-net` gives the error 
      ```[KNL] allocating SPI failed: Operation not permitted (1)```

```console
moon@moon:~$ swanctl --initiate --child net-net
[IKE] initiating IKE_SA site-site[4] to 192.168.57.7
[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
[NET] sending packet: from 192.168.57.3[500] to 192.168.57.7[500] (604 bytes)
[NET] received packet: from 192.168.57.7[500] to 192.168.57.3[500] (248 bytes)
[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
[CFG] selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/CURVE_25519
[IKE] authentication of 'moon.strongswan.org' (myself) with ED25519 successful
[IKE] establishing CHILD_SA net-net{5}
[KNL] allocating SPI failed: Operation not permitted (1)
[KNL] unable to get SPI
[IKE] unable to allocate SPIs from kernel
initiate failed: establishing CHILD_SA 'net-net' failed
```

I read [issue 996](https://wiki.strongswan.org/issues/996) and understood that I have to set some capabilities to the user (or to the executable?) who start the daemon.

Would you please give me some hint? I have read the "Reduced Privileges" page of the documentation, but some part are not very clear to me and that is why I am asking here.

Thanks in advance

**Reply by tobiasbrunner** [reply]
> Edited the /etc/strongswan.conf file and added the row user=myuser to be able to execute the swanctl tool as non-root

Note that you don't have to configure a user (or your own user) for that, you could achieve the same also if the daemon continues to run as root (or some other service user) by configuring `group` and assigning users that are allowed to access the vici UNIX socket to that group.

> Would you please give me some hint?

Either start the daemon as root user (it will drop capabilities and change to the configured user after it opened privileged sockets etc.), or add the required capabilities to the charon/charon-systemd executable via `setcap(8)` and let the a non-root user start it directly (`user` has to be set to that user, or not configured at all).

**Reply by angelofloridia** [reply]
Thanks a lot! 
For me of the future and for anyone else with this situation, I managed to add the correct privileges by 
```
sudo setcap cap_net_admin=eip /usr/sbin/charon-systemd
```
and leaving the user configured in the `strongswan.conf` file.

Is this a decent way of using it?

**Reply by tobiasbrunner** [reply]
Depends on what your goal is (it basically allows any user to start the daemon). But I wonder why setting the capability on the executable is actually necessary. Doesn't `systemd` start the daemon as root user? Or did you change that as well? Maybe capability dropping (or rather retaining the required capabilities) doesn't work correctly. Are any errors logged when the daemon is starting, or does it say `"dropped capabilities, running as ..."`?

**Reply by angelofloridia** [reply]
I'm writing a Java code that is a piece of a larger framework. The goal is to establish vpn tunnels based on just the ip using scp and ssh. To be able to activate the tunnel using the Java library `jsch` I cannot use `sudo`, but I can use the user.

To answer your question, I start the daemon as sudo (`sudo systemctl start strongswan` and nothing about dropped capabilities is shown in terminal.

Super stupid question, where can I see the logs?

Thank you.

**Reply by tobiasbrunner** [reply]
> I'm writing a Java code that is a piece of a larger framework. The goal is to establish vpn tunnels based on just the ip using scp and ssh. To be able to activate the tunnel using the Java library jsch I cannot use sudo, but I can use the user.

Not sure I understand this completely (e.g. how IPsec plays a role in it or the interaction between jsch and strongSwan). Also, you might want to have a look at [vici](https://docs.strongswan.org/docs/5.9/plugins/vici.html), and maybe initiation via trap policies would also work.

>  I start the daemon as sudo (sudo systemctl start strongswan and nothing about dropped capabilities is shown in terminal.

Then I suppose it's started as root. Check the log for corresponding log messages.

> Super stupid question, where can I see the logs?

Depends on the daemon and the [logging configuration](https://docs.strongswan.org/docs/5.9/config/logging.html), but with ` charon-systemd` and the default config e.g. via `journalctl -u strongswan`.

**Reply by angelofloridia** [reply]
Sorry, maybe I was too brief: I have an arbitrary number of virtual machines, each having its user. I need to remotely instantiate tunnels by pushing swanctl.conf files in the right machines and then setup the tunnel.

I already use `start_action = trap|start` but most of the times it does not work (I am missing something for sure) and so I wrote the java code to execute `swanctl --load-all` after pushing the swanctl.conf file. This last command I couldn't  execute as root, that's why I wanted to execute it as user.

From the logs I can see this after restarting the service: `moon charon-systemd[3325]: dropped capabilities, running as uid 1000, gid 0` so it drops capabilities.

**Reply by tobiasbrunner** [reply]
> From the logs I can see this after restarting the service: `moon charon-systemd[3325]: dropped capabilities, running as uid 1000, gid 0` so it drops capabilities.

Hm, then `setcap cap_net_admin=eip` shouldn't be necessary as the NET_ADMIN capability should be retained by `capset()` (with `native` interface, with `libcap` by `cap_set_proc()`) right before that log message.

**Reply by angelofloridia** [reply]
> Hm, then `setcap cap_net_admin=eip` shouldn't be necessary as the NET_ADMIN capability should be retained by `capset()` (with `native` interface, with `libcap` by `cap_set_proc()`) right before that log message.

you're right! In fact I removed the capability from the executable and it still works as needed (I use `native` interface)

This is the log starting the daemon (I know it missed the swanctl.conf file):
```console
Apr 22 16:10:50 moon systemd[1]: Starting strongSwan IPsec IKEv1/IKEv2 daemon using swanctl...
Apr 22 16:10:50 moon charon-systemd[3878]: install DNS servers in '/etc/resolv.conf'
Apr 22 16:10:50 moon charon-systemd[3878]: loading ca certificates from '/etc/ipsec.d/cacerts'
Apr 22 16:10:50 moon charon-systemd[3878]: loading aa certificates from '/etc/ipsec.d/aacerts'
Apr 22 16:10:50 moon charon-systemd[3878]: loading ocsp signer certificates from '/etc/ipsec.d/ocspcerts'
Apr 22 16:10:50 moon charon-systemd[3878]: loading attribute certificates from '/etc/ipsec.d/acerts'
Apr 22 16:10:50 moon charon-systemd[3878]: loading crls from '/etc/ipsec.d/crls'
Apr 22 16:10:50 moon charon-systemd[3878]: loading secrets from '/etc/ipsec.secrets'
Apr 22 16:10:50 moon charon-systemd[3878]: loaded plugins: charon-systemd attr cmac constraints counters des dnskey drbg fips-prf gcm md5 mgf1 pgp pkcs12 pkcs7 pkcs8 pubkey random rc2 resolve socket-default sshkey stroke updown vic
i xauth-generic xcbc nonce aes sha1 sha2 hmac kdf pem pkcs1 x509 revocation curve25519 gmp kernel-netlink
Apr 22 16:10:50 moon charon-systemd[3878]: dropped capabilities, running as uid 1000, gid 0
Apr 22 16:10:50 moon charon-systemd[3878]: spawning 16 worker threads
Apr 22 16:10:50 moon swanctl[3895]: no files found matching '/etc/swanctl/swanctl.conf'
Apr 22 16:10:50 moon swanctl[3895]: failed to open config file '/etc/swanctl/swanctl.conf'
Apr 22 16:10:50 moon charon-systemd[3878]: loaded certificate 'C=CH, O=strongswan, CN=sun.strongswan.org'
Apr 22 16:10:50 moon charon-systemd[3878]: loaded certificate 'C=CH, O=strongswan, CN=moon.strongswan.org'
Apr 22 16:10:50 moon charon-systemd[3878]: loaded certificate 'C=CH, O=strongSwan, CN=strongSwan Root CA'
Apr 22 16:10:50 moon charon-systemd[3878]: loaded ANY private key
Apr 22 16:10:50 moon swanctl[3895]: loaded certificate from '/etc/swanctl/x509/sunCert.pem'
Apr 22 16:10:50 moon swanctl[3895]: loaded certificate from '/etc/swanctl/x509/moonCert.pem'
Apr 22 16:10:50 moon swanctl[3895]: loaded certificate from '/etc/swanctl/x509ca/strongswanCert.pem'
Apr 22 16:10:50 moon swanctl[3895]: loaded private key from '/etc/swanctl/private/moonKey.pem'
Apr 22 16:10:50 moon swanctl[3895]: no authorities found, 0 unloaded
Apr 22 16:10:50 moon swanctl[3895]: no pools found, 0 unloaded
Apr 22 16:10:50 moon swanctl[3895]: no connections found, 0 unloaded
Apr 22 16:10:50 moon systemd[1]: Started strongSwan IPsec IKEv1/IKEv2 daemon using swanctl.
```

Everything seems to be ok to me.

Anyway I'll share the work as I finish, it's my master thesis.
Thank you for the support given till now, I appreciate it.
